File: src/nexora/web/internal/contracts.py
"""Stable internal frontend response contracts."""

from datetime import datetime

from pydantic import BaseModel, Field


class InternalError(BaseModel):
    code: str
    message: str
    field_errors: dict[str, list[str]] = Field(default_factory=dict)
    conflict: dict[str, object] | None = None
    task_id: str | None = None


class AdministratorPreferences(BaseModel):
    username: str
    global_monospace: bool
    density: str
    language: str
    timezone: str


class InternalSession(BaseModel):
    administrator: AdministratorPreferences
    csrf_token: str
    features: dict[str, bool]


class OverviewSummary(BaseModel):
    host_total: int
    host_ready: int
    host_synced: int
    vm_total: int
    vm_running: int
    vm_paused: int
    vm_stopped: int
    active_tasks: int
    task_pending: int
    failed_tasks: int
    storage_pool_total: int
    storage_volume_total: int


class HostSummary(BaseModel):
    id: str
    name: str
    address: str
    ssh_port: int
    status: str
    labels: list[str]
    last_scanned_at: datetime | None


class VmSummary(BaseModel):
    resource_id: str
    host_id: str
    native_id: str
    name: str
    host_name: str
    state: str
    status: str
    vcpus: int | None
    memory_mib: int | None
    last_seen_at: datetime
    needs_restart: bool = False


class HostListResponse(BaseModel):
    items: list[HostSummary]
    total: int
    page: int
    page_size: int


class VmListResponse(BaseModel):
    items: list[VmSummary]
    total: int
    page: int
    page_size: int


class HostFeatureSummary(BaseModel):
    key: str
    name: str
    status: str
    description: str


class HostHardwareSummary(BaseModel):
    manufacturer: str | None
    model: str | None
    os_name: str | None
    kernel: str | None
    architecture: str | None
    cpu_model: str | None
    logical_cpus: int | None
    sockets: int | None
    cores_per_socket: int | None
    threads_per_core: int | None
    numa_nodes: int | None
    memory_bytes: int | None


class HostNetworkAdapterSummary(BaseModel):
    name: str
    mac: str | None
    kind: str
    state: str
    management: bool


class HostMetricSummary(BaseModel):
    sampled_at: datetime
    load_1: float
    load_5: float
    load_15: float
    memory_total_kib: int
    memory_available_kib: int
    uptime_seconds: int


class HostDetailResponse(BaseModel):
    host: HostSummary
    ssh_username: str
    libvirt_uri: str
    resource_counts: dict[str, int]
    virtual_machines: list[VmSummary]
    features: list[HostFeatureSummary]
    hardware: HostHardwareSummary
    network_adapters: list[HostNetworkAdapterSummary]
    latest_metrics: HostMetricSummary | None
    manage_url: str


class VmDiskSummary(BaseModel):
    type: str | None
    device: str | None
    source: str | None
    target: str | None
    bus: str | None
    format: str | None
    readonly: bool
    shareable: bool


class VmInterfaceSummary(BaseModel):
    type: str | None
    source: str | None
    mac: str | None
    target: str | None
    model: str | None


class VmHostDeviceSummary(BaseModel):
    type: str
    address: str
    category: str
    name: str
    driver: str | None
    iommu_group: str | None


class VmSnapshotSummary(BaseModel):
    resource_id: str
    name: str
    status: str
    state: str
    creation_time: str | None
    current: bool
    memory: str | None
    disks: list[str]
    xml: str | None


class VmMetricSummary(BaseModel):
    sampled_at: datetime
    state: str
    cpu_usage_percent: float | None
    memory_usage_kib: int | None
    disk_read_bytes: int | None
    disk_write_bytes: int | None
    net_rx_bytes: int | None
    net_tx_bytes: int | None


class VmDetailResponse(BaseModel):
    vm: VmSummary
    active: bool
    persistent: bool
    autostart: bool
    maximum_vcpus: int | None
    configuration_status: str
    disks: list[VmDiskSummary]
    interfaces: list[VmInterfaceSummary]
    host_devices: list[VmHostDeviceSummary]
    snapshots: list[VmSnapshotSummary]
    metrics: list[VmMetricSummary]
    xml: str
    manage_url: str


class GuestAddressSummary(BaseModel):
    interface: str
    address: str
    prefix: int
    family: str
    mac: str | None


class GuestAgentResponse(BaseModel):
    state: str
    channel_configured: bool
    hostname: str | None
    addresses: list[GuestAddressSummary]
    message: str | None


class VmCreateVolumeOption(BaseModel):
    id: str
    host_id: str
    host_name: str
    pool_name: str
    name: str
    format: str
    capacity_bytes: int


class VmCreateNetworkOption(BaseModel):
    id: str
    host_id: str
    host_name: str
    kind: str
    name: str


class VmCreateIsoOption(BaseModel):
    id: str
    host_id: str
    host_name: str
    pool_name: str
    name: str


class VmCreateOptionsResponse(BaseModel):
    volumes: list[VmCreateVolumeOption]
    networks: list[VmCreateNetworkOption]
    isos: list[VmCreateIsoOption]


class VmCreatePreviewRequest(BaseModel):
    name: str = Field(min_length=1, max_length=128)
    memory_mib: int = Field(ge=128, le=1_048_576)
    vcpus: int = Field(ge=1, le=4_096)
    volume_resource_id: str
    network_resource_id: str = ""
    iso_resource_id: str = ""
    driver_iso_resource_id: str = ""
    disk_bus: str = "virtio"
    cpu_mode: str = "host-model"
    guest_profile: str = "linux"
    firmware: str = "bios"
    secure_boot: bool = False
    tpm2: bool = False


class VmCreatePreviewSummary(BaseModel):
    name: str
    host_name: str
    vcpus: int
    memory_mib: int
    volume_name: str
    network: str
    iso_name: str | None
    firmware: str
    secure_boot: bool
    tpm2: bool


class VmCreatePreviewResponse(BaseModel):
    plan_id: str
    confirmation_token: str
    host_id: str
    vm_uuid: str
    diff_text: str
    summary: VmCreatePreviewSummary


class VmCreateApplyRequest(BaseModel):
    plan_id: str
    confirmation_token: str
    host_id: str
    vm_uuid: str


class VmBlankCreatePoolOption(BaseModel):
    id: str
    host_id: str
    host_name: str
    pool_name: str
    target_path: str


class VmBlankCreateOptionsResponse(BaseModel):
    pools: list[VmBlankCreatePoolOption]
    networks: list[VmCreateNetworkOption]
    isos: list[VmCreateIsoOption]


class VmBlankCreatePreviewRequest(BaseModel):
    name: str = Field(min_length=1, max_length=128)
    memory_mib: int = Field(ge=128, le=1_048_576)
    vcpus: int = Field(ge=1, le=4_096)
    pool_resource_id: str
    disk_name: str = Field(min_length=1, max_length=255)
    volume_format: str = "qcow2"
    capacity_gib: int = Field(ge=1, le=8192)
    network_resource_id: str = ""
    iso_resource_id: str = ""
    driver_iso_resource_id: str = ""
    disk_bus: str = "virtio"
    cpu_mode: str = "host-model"
    guest_profile: str = "linux"
    firmware: str = "bios"
    secure_boot: bool = False
    tpm2: bool = False


class VmBlankCreatePreviewSummary(BaseModel):
    name: str
    host_name: str
    vcpus: int
    memory_mib: int
    pool_name: str
    disk_name: str
    volume_format: str
    capacity_bytes: int
    network: str
    iso_name: str | None
    firmware: str
    secure_boot: bool
    tpm2: bool


class VmBlankCreatePreviewResponse(BaseModel):
    plan_id: str
    confirmation_token: str
    host_id: str
    vm_uuid: str
    diff_text: str
    summary: VmBlankCreatePreviewSummary


class TaskCreatedResponse(BaseModel):
    task_id: str
    location: str


class HostOnboardingRequest(BaseModel):
    name: str = Field(min_length=1, max_length=128)
    address: str = Field(min_length=1, max_length=255)
    ssh_port: int = Field(ge=1, le=65_535)
    ssh_username: str = Field(min_length=1, max_length=64)
    authentication_method: str
    password: str | None = None
    private_key: str | None = None
    private_key_passphrase: str | None = None
    sudo_mode: str
    labels: list[str] = Field(default_factory=list, max_length=32)
    notes: str | None = Field(default=None, max_length=4_000)


class HostKeySummary(BaseModel):
    key_type: str
    fingerprint: str


class HostOnboardingStartedResponse(BaseModel):
    host_id: str
    location: str


class HostKeyConfirmationResponse(BaseModel):
    host_id: str
    name: str
    endpoint: str
    host_key_digest: str
    fingerprints: list[HostKeySummary]


class HostKeyConfirmRequest(BaseModel):
    host_key_digest: str


class VmMediaOption(BaseModel):
    id: str
    file_name: str
    format: str
    size_bytes: int
    virtual_size_bytes: int | None


class VmMediaTargetOption(BaseModel):
    id: str
    host_id: str
    host_name: str
    pool_name: str
    target_path: str


class VmMediaCreateOptionsResponse(BaseModel):
    media: list[VmMediaOption]
    targets: list[VmMediaTargetOption]
    networks: list[VmCreateNetworkOption]
    isos: list[VmCreateIsoOption]


class VmMediaCreatePreviewRequest(BaseModel):
    name: str = Field(min_length=1, max_length=128)
    memory_mib: int = Field(ge=128, le=1_048_576)
    vcpus: int = Field(ge=1, le=4_096)
    media_item_id: str
    pool_resource_id: str
    target_file_name: str = Field(min_length=1, max_length=255)
    target_capacity_gib: int | None = Field(default=None, ge=1, le=16_384)
    network_resource_id: str = ""
    iso_resource_id: str = ""
    driver_iso_resource_id: str = ""
    disk_bus: str = "virtio"
    cpu_mode: str = "host-model"
    guest_profile: str = "linux"
    firmware: str = "bios"
    secure_boot: bool = False
    tpm2: bool = False
    cloud_init: bool = False
    cloud_hostname: str = ""
    cloud_username: str = ""
    cloud_ssh_public_key: str = ""
    cloud_password: str = ""
    cloud_password_confirmation: str = ""
    cloud_network_mode: str = "dhcp"
    cloud_ipv4_cidr: str = ""
    cloud_ipv4_gateway: str = ""
    cloud_ipv6_cidr: str = ""
    cloud_ipv6_gateway: str = ""
    cloud_dns_addresses: list[str] = Field(default_factory=list, max_length=3)


class VmMediaCreatePreviewSummary(BaseModel):
    name: str
    host_name: str
    media_name: str
    pool_name: str
    target_path: str
    source_sha256: str
    vcpus: int
    memory_mib: int
    target_capacity_bytes: int | None
    network: str
    iso_name: str | None
    driver_iso_name: str | None
    firmware: str
    secure_boot: bool
    tpm2: bool
    cloud_init: str | None


class VmMediaCreatePreviewResponse(BaseModel):
    plan_id: str
    confirmation_token: str
    host_id: str
    vm_uuid: str
    diff_text: str
    summary: VmMediaCreatePreviewSummary

File: frontend/src/api/contracts.ts
export type AdministratorPreferences = {
  username: string;
  global_monospace: boolean;
  density: string;
  language: string;
  timezone: string;
};

export type InternalSession = {
  administrator: AdministratorPreferences;
  csrf_token: string;
  features: Record<string, boolean>;
};

export type InternalErrorPayload = {
  code: string;
  message: string;
  field_errors: Record<string, string[]>;
  conflict: Record<string, unknown> | null;
  task_id: string | null;
};

export type OverviewSummary = {
  host_total: number;
  host_ready: number;
  host_synced: number;
  vm_total: number;
  vm_running: number;
  vm_paused: number;
  vm_stopped: number;
  active_tasks: number;
  task_pending: number;
  failed_tasks: number;
  storage_pool_total: number;
  storage_volume_total: number;
};

export type HostSummary = {
  id: string;
  name: string;
  address: string;
  ssh_port: number;
  status: string;
  labels: string[];
  last_scanned_at: string | null;
};

export type VmSummary = {
  resource_id: string;
  host_id: string;
  native_id: string;
  name: string;
  host_name: string;
  state: string;
  status: string;
  vcpus: number | null;
  memory_mib: number | null;
  last_seen_at: string;
  needs_restart: boolean;
};

export type PaginatedResponse<T> = {
  items: T[];
  total: number;
  page: number;
  page_size: number;
};

export type HostFeatureSummary = {
  key: string;
  name: string;
  status: string;
  description: string;
};

export type HostHardwareSummary = {
  manufacturer: string | null;
  model: string | null;
  os_name: string | null;
  kernel: string | null;
  architecture: string | null;
  cpu_model: string | null;
  logical_cpus: number | null;
  sockets: number | null;
  cores_per_socket: number | null;
  threads_per_core: number | null;
  numa_nodes: number | null;
  memory_bytes: number | null;
};

export type HostNetworkAdapterSummary = {
  name: string;
  mac: string | null;
  kind: string;
  state: string;
  management: boolean;
};

export type HostMetricSummary = {
  sampled_at: string;
  load_1: number;
  load_5: number;
  load_15: number;
  memory_total_kib: number;
  memory_available_kib: number;
  uptime_seconds: number;
};

export type HostDetail = {
  host: HostSummary;
  ssh_username: string;
  libvirt_uri: string;
  resource_counts: Record<string, number>;
  virtual_machines: VmSummary[];
  features: HostFeatureSummary[];
  hardware: HostHardwareSummary;
  network_adapters: HostNetworkAdapterSummary[];
  latest_metrics: HostMetricSummary | null;
  manage_url: string;
};

export type VmDiskSummary = {
  type: string | null;
  device: string | null;
  source: string | null;
  target: string | null;
  bus: string | null;
  format: string | null;
  readonly: boolean;
  shareable: boolean;
};

export type VmInterfaceSummary = {
  type: string | null;
  source: string | null;
  mac: string | null;
  target: string | null;
  model: string | null;
};

export type VmHostDeviceSummary = {
  type: string;
  address: string;
  category: string;
  name: string;
  driver: string | null;
  iommu_group: string | null;
};

export type VmSnapshotSummary = {
  resource_id: string;
  name: string;
  status: string;
  state: string;
  creation_time: string | null;
  current: boolean;
  memory: string | null;
  disks: string[];
  xml: string | null;
};

export type VmMetricSummary = {
  sampled_at: string;
  state: string;
  cpu_usage_percent: number | null;
  memory_usage_kib: number | null;
  disk_read_bytes: number | null;
  disk_write_bytes: number | null;
  net_rx_bytes: number | null;
  net_tx_bytes: number | null;
};

export type VmDetail = {
  vm: VmSummary;
  active: boolean;
  persistent: boolean;
  autostart: boolean;
  maximum_vcpus: number | null;
  configuration_status: string;
  disks: VmDiskSummary[];
  interfaces: VmInterfaceSummary[];
  host_devices: VmHostDeviceSummary[];
  snapshots: VmSnapshotSummary[];
  metrics: VmMetricSummary[];
  xml: string;
  manage_url: string;
};

export type VmLifecycleAction =
  | "start"
  | "shutdown"
  | "force_off"
  | "reboot"
  | "force_reboot"
  | "pause"
  | "resume"
  | "managed_save"
  | "autostart_enable"
  | "autostart_disable";

export type TaskSummary = {
  id: string;
  title: string;
  task_type: string;
  status: string;
  progress: number;
  current_step: number;
  total_steps: number;
  message: string | null;
  error_message: string | null;
  created_at: string;
  started_at: string | null;
  finished_at: string | null;
  resumable: boolean;
  retry_count: number;
  max_retries: number;
  host_id: string | null;
  host_name: string | null;
};

export type TaskStepSummary = {
  sequence: number;
  name: string;
  status: string;
  attempt_count: number;
  started_at: string | null;
  finished_at: string | null;
  error_code: string | null;
};

export type TaskDetail = { task: TaskSummary; steps: TaskStepSummary[] };

export type MediaItemSummary = {
  id: string;
  relative_path: string;
  file_name: string;
  kind: string;
  status: string;
  size_bytes: number;
  sha256: string;
  image_format: string | null;
  classification: string | null;
  architecture: string | null;
  standalone: boolean;
};

export type GuestAgentSummary = {
  state: string;
  channel_configured: boolean;
  hostname: string | null;
  addresses: Array<{
    interface: string;
    address: string;
    prefix: number;
    family: string;
    mac: string | null;
  }>;
  message: string | null;
};

export type VmCreateVolumeOption = {
  id: string;
  host_id: string;
  host_name: string;
  pool_name: string;
  name: string;
  format: string;
  capacity_bytes: number;
};

export type VmCreateNetworkOption = {
  id: string;
  host_id: string;
  host_name: string;
  kind: string;
  name: string;
};

export type VmCreateIsoOption = {
  id: string;
  host_id: string;
  host_name: string;
  pool_name: string;
  name: string;
};

export type VmCreateOptions = {
  volumes: VmCreateVolumeOption[];
  networks: VmCreateNetworkOption[];
  isos: VmCreateIsoOption[];
};

export type VmCreatePreviewRequest = {
  name: string;
  memory_mib: number;
  vcpus: number;
  volume_resource_id: string;
  network_resource_id: string;
  iso_resource_id: string;
  driver_iso_resource_id: string;
  disk_bus: string;
  cpu_mode: string;
  guest_profile: string;
  firmware: string;
  secure_boot: boolean;
  tpm2: boolean;
};

export type VmCreatePreview = {
  plan_id: string;
  confirmation_token: string;
  host_id: string;
  vm_uuid: string;
  diff_text: string;
  summary: {
    name: string;
    host_name: string;
    vcpus: number;
    memory_mib: number;
    volume_name: string;
    network: string;
    iso_name: string | null;
    firmware: string;
    secure_boot: boolean;
    tpm2: boolean;
  };
};

export type VmBlankCreatePoolOption = {
  id: string;
  host_id: string;
  host_name: string;
  pool_name: string;
  target_path: string;
};

export type VmBlankCreateOptions = {
  pools: VmBlankCreatePoolOption[];
  networks: VmCreateNetworkOption[];
  isos: VmCreateIsoOption[];
};

export type VmBlankCreatePreviewRequest = {
  name: string;
  memory_mib: number;
  vcpus: number;
  pool_resource_id: string;
  disk_name: string;
  volume_format: string;
  capacity_gib: number;
  network_resource_id: string;
  iso_resource_id: string;
  driver_iso_resource_id: string;
  disk_bus: string;
  cpu_mode: string;
  guest_profile: string;
  firmware: string;
  secure_boot: boolean;
  tpm2: boolean;
};

export type VmBlankCreatePreview = {
  plan_id: string;
  confirmation_token: string;
  host_id: string;
  vm_uuid: string;
  diff_text: string;
  summary: {
    name: string;
    host_name: string;
    vcpus: number;
    memory_mib: number;
    pool_name: string;
    disk_name: string;
    volume_format: string;
    capacity_bytes: number;
    network: string;
    iso_name: string | null;
    firmware: string;
    secure_boot: boolean;
    tpm2: boolean;
  };
};

export type VmRemovePreview = {
  plan_id: string;
  confirmation_token: string;
  operation: string;
  target_name: string | null;
  remove_disks: boolean;
  remove_nvram: boolean;
  diff_text: string;
};

export type TaskCreated = { task_id: string; location: string };

export type HostOnboardingRequest = {
  name: string;
  address: string;
  ssh_port: number;
  ssh_username: string;
  authentication_method: string;
  password: string | null;
  private_key: string | null;
  private_key_passphrase: string | null;
  sudo_mode: string;
  labels: string[];
  notes: string | null;
};

export type HostOnboardingStarted = { host_id: string; location: string };

export type HostKeyConfirmation = {
  host_id: string;
  name: string;
  endpoint: string;
  host_key_digest: string;
  fingerprints: Array<{ key_type: string; fingerprint: string }>;
};

export type VmMediaCreateOptions = {
  media: Array<{ id: string; file_name: string; format: string; size_bytes: number; virtual_size_bytes: number | null }>;
  targets: Array<{ id: string; host_id: string; host_name: string; pool_name: string; target_path: string }>;
  networks: VmCreateNetworkOption[];
  isos: VmCreateIsoOption[];
};

export type VmMediaCreatePreviewRequest = {
  name: string;
  memory_mib: number;
  vcpus: number;
  media_item_id: string;
  pool_resource_id: string;
  target_file_name: string;
  target_capacity_gib: number | null;
  network_resource_id: string;
  iso_resource_id: string;
  driver_iso_resource_id: string;
  disk_bus: string;
  cpu_mode: string;
  guest_profile: string;
  firmware: string;
  secure_boot: boolean;
  tpm2: boolean;
  cloud_init: boolean;
  cloud_hostname: string;
  cloud_username: string;
  cloud_ssh_public_key: string;
  cloud_password: string;
  cloud_password_confirmation: string;
  cloud_network_mode: string;
  cloud_ipv4_cidr: string;
  cloud_ipv4_gateway: string;
  cloud_ipv6_cidr: string;
  cloud_ipv6_gateway: string;
  cloud_dns_addresses: string[];
};

export type VmMediaCreatePreview = {
  plan_id: string;
  confirmation_token: string;
  host_id: string;
  vm_uuid: string;
  diff_text: string;
  summary: {
    name: string;
    host_name: string;
    media_name: string;
    pool_name: string;
    target_path: string;
    source_sha256: string;
    vcpus: number;
    memory_mib: number;
    target_capacity_bytes: number | null;
    network: string;
    iso_name: string | null;
    driver_iso_name: string | null;
    firmware: string;
    secure_boot: boolean;
    tpm2: boolean;
    cloud_init: string | null;
  };
};

export type StorageOverview = {
  hosts: Array<{ id: string; name: string }>;
  pools: StoragePoolSummary[];
  volumes: StorageVolumeSummary[];
};

export type StoragePoolSummary = {
  resource_id: string;
  host_id: string;
  host_name: string;
  native_id: string;
  name: string;
  status: string;
  pool_type: string;
  state: string;
  active: boolean;
  autostart: boolean;
  target_path: string | null;
  capacity_bytes: number | null;
  available_bytes: number | null;
  writable: boolean;
};

export type StorageVolumeSummary = {
  resource_id: string;
  host_id: string;
  host_name: string;
  pool_resource_id: string;
  pool_name: string;
  name: string;
  status: string;
  format: string | null;
  capacity_bytes: number | null;
  allocation_bytes: number | null;
  in_use: boolean;
  writable: boolean;
};

export type StorageChangePreview = {
  plan_id: string;
  confirmation_token: string;
  host_id: string;
  pool_uuid: string;
  diff_text: string;
  operation: string;
  summary: Record<string, unknown>;
};

export type StoragePoolCreateRequest = {
  host_id: string;
  name: string;
  pool_type: string;
  target_path: string;
  source_host: string | null;
  source_path: string | null;
  nfs_version: string | null;
  mount_options: string[];
  start: boolean;
  autostart: boolean;
};

export type StorageVolumeCreateRequest = {
  pool_resource_id: string;
  name: string;
  volume_format: string;
  capacity_gib: number;
};
